enhancements/ingress/header-case-adjustment.md

```diff
@@ -0,0 +1,335 @@
+specified.
+
+### Implementation Details
+
+Implementing this enhancement required changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+* openshift/router
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  The enhancement adds a new
+environment variable, `ROUTER_H1_CASE_ADJUST`, which specifies adjustments as a
+comma-delimited list of header names in the desired capitalizations.  The router
+parses the value of `ROUTER_H1_CASE_ADJUST` and translates it into
+`h1-case-adjust` settings.  If `ROUTER_H1_CASE_ADJUST` is nonempty, the router
+also specifies the `h1-case-adjust-bogus-server` setting.
+
+For example, if `ROUTER_H1_CASE_ADJUST` is set to `Host,X-Forwarded-For`, then
+the router specifies the following settings in the HAProxy configuration:
+
+```
+global
+  # ...
+  h1-case-adjust host Host
+  h1-case-adjust x-forwarded-for X-Forwarded-For
+  # ...
+
+defaults
+  # ...
+  option h1-case-adjust-bogus-client
+  # ...
+```
+
+When a Route has the `haproxy.router.openshift.io/h1-adjust-case=true`
+annotation, the router specifies the `h1-case-adjust-bogus-server` setting on
+the configuration for that Route:
+
+```
+backend ...
+  option h1-case-adjust-bogus-server
+```
+
+The ingress operator translates the value of the new
+`spec.httpHeaders.headerNameCaseAdjustments` API field into the format for the
+`ROUTER_H1_CASE_ADJUST` environment variable and sets this environment variable
+to that value in the router Deployment.
+
+### Risks and Mitigations
+
+OpenShift cannot infer which HTTP header names legacy clients and servers expect
+to be capitalized a certain way, so cluster administrators and application
+developers need to be aware of the issue and configure IngressControllers and
+Routes using this new configuration when updating to OpenShift 4.7.  A release
+note may mitigate this concern; for example:
+
+"OpenShift 4.7 updates to HAProxy 2.2, which down-cases HTTP header names by
+default (for example, `Host: xyz.com` is transformed to `host: xyz.com`).  If
+legacy applications are sensitive to the capitalization of HTTP header names,
+see the IngressController's new `spec.httpHeaders.headerNameCaseAdjustments` API
+field for a solution to accommodate these legacy applications until they can be
+fixed."
+
+## Design Details
+
+### Test Plan
+
+The controller that manages the router Deployment and related resources has unit
+test coverage; for this enhancement, the unit tests have been expanded to cover
+the additional functionality.
+
+The operator has end-to-end tests; for this enhancement, a test has been added
+that (1) creates an IngressController configured with case adjustments, (2)
+creates a Pod and Route for an HTTP application that echoes back requests, (3)
+sends a request to the Route, and (4) verifies that the application echoes back
+the expected headers.
```

Comment:
nit: the test should also verify that the response itself sets the correct header cases in it's own headers, right?

Reply:
Right, it checks both.  I'll make that clearer.